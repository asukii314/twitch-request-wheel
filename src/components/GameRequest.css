.game-request {
    box-shadow: inset 1px 1px 40px 0 rgba(0, 0, 0, 0.45);
    background-size: cover;
    border-radius: 5px;
    text-align: left;
    text-transform: capitalize;
}

.game-request.chosen {
    background-color: greenyellow;
    color: black;
}

.game-request.pending {
    background-color: steelblue;
}

.game-request-body {
    align-items: center;
    display: flex;
    font-size: 18px;
    letter-spacing: -0.3px;
    justify-content: space-between;
    margin: 0px 15px 7px;
    padding: 2px;
}

.game-request-body .options {
    display: flex;
}

.game-request .lock {
    width: 24px;
    opacity:  0.2;
    padding-right: 8px;
    position: absolute;
    margin-left: -24px;
    margin-top: 3px;
}

.game-request .deleteButton {
    background-color: red;
    float: right;
    height: 50%;
    color: white;
    font-size: 11px;
    opacity: 0.5;
    border-radius: 1rem;
}
.game-request .lock,
.game-request .deleteButton {
    cursor: pointer;
}
.game-request .lock:hover,
.game-request .lock:active,
.game-request .lock.locked,
.game-request .deleteButton:hover,
.game-request .deleteButton:active {
    opacity: 1;
}


#game-request-tooltip .tooltip-inner {
    max-width: 500px;
}

.game-request .tooltip {
    display: flex;
    align-items: center;
}

.game-request .tooltipText {
    margin: 0;
}

#game-request-tooltip .status {
    display: inline-block;
    width: 8px;
    height: 8px;
    border-radius: 5px;
    margin-left: 6px;
    margin-top: 3px;
}

#game-request-tooltip .status.active {
    background-color: green;
}

#game-request-tooltip .status.idle {
    background-color: yellow;
}

#game-request-tooltip .status.disconnected {
    background-color: darkgrey;
}
