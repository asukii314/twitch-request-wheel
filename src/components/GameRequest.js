import {ActivityStatus} from '../ChatActivity';
import React, {Component} from 'react';
import {OverlayTrigger, Tooltip} from 'react-bootstrap'
import lock from '../images/lock.svg';

import './GameRequest.css';

class GameRequest extends Component {
    constructor(props) {
        super(props);
        this.state = {
            timeDiff: 0,
            activityStatus: null
        };

        this.deleteRequest = this.deleteRequest.bind(this);
        this.getFormattedTimeDiff = this.getFormattedTimeDiff.bind(this);
        this.toggleLock = this.toggleLock.bind(this);
        this.updateStatus = this.updateStatus.bind(this);
    }

    deleteRequest = (e) => {
        this.props.onDelete(this.props.gameName)
    }

    getFormattedTimeDiff = (timestamp) => {
        let unit = "minute";
        let timeDiff = Math.floor((Date.now() - timestamp) / 60000);
        if (timeDiff === 0) {
            return "just now";
        }
        if (timeDiff >= 60) {
            unit = "hour";
            timeDiff = Math.floor(timeDiff / 60);

            if (timeDiff >= 24) {
                unit = "day";
                timeDiff = Math.floor(timeDiff / 24);
            }
        }
        return `${timeDiff} ${unit}${timeDiff === 1 ? "" : "s"} ago`;
    };

    getTooltipContent = (props) => (
        <Tooltip id="game-request-tooltip" {...props}>
            <span className="tooltipText">
                Requested {this.state.timeDiff} by {this.props.metadata.username}
                <span className={`status ${this.getTooltipActivityStatusClassName()}`} />
            </span>
        </Tooltip>
    );

    getTooltipActivityStatusClassName = () => {
        switch (this.state.activityStatus) {
            case ActivityStatus.ACTIVE:
                return "active";
            case ActivityStatus.IDLE:
                return "idle";
            case ActivityStatus.DISCONNECTED:
                return "disconnected";
            default:
                // no data back yet; don't show an activity status indicator at all
                return '';
        }
    }

    toggleLock = () => {
        return this.props.toggleLock(this.props.gameName);
    }

    updateStatus = async () => {
        // return Date.now();
        let activityStatus;
        try {
            activityStatus = await this.props.getActivity(this.props.metadata.username);
            return this.setState({
                timeDiff: this.getFormattedTimeDiff(this.props.metadata.time),
                activityStatus
            });
        } catch(err) {
            throw err;
        }
    }

    render() {
        const lockClassName = this.props.metadata.locked ? 'lock locked' : 'lock unlocked';
        const cardStatus = this.props.metadata.chosen ? 'chosen' : 'pending';

        return (
            <OverlayTrigger
                placement="left"
                overlay={this.getTooltipContent}
                onEnter={this.updateStatus}>
                <div className="game-request-wrapper fade-in">
                	<div id="baseDiv" className={`game-request ${cardStatus}`}>
                		<div className="game-request-body">
                		    <div className="text-truncate">
                                <b>
                                    {this.props.metadata.name}
                                </b>
                                {' '}
                                <small>
                                    {
                                        !!this.props.metadata.partyPack &&
                                        `(${this.props.metadata.partyPack.replace('Party Pack', 'Pack')})`
                                    }
                                </small>
                            </div>
                			<div className="options">
                				<img src={lock} alt="lock" className={lockClassName} onClick={this.toggleLock} />
                				<button type='button' className="deleteButton" onClick={this.deleteRequest}>X</button>
                			</div>
                		</div>
                	</div>
                </div>
            </OverlayTrigger>
        );
    }
}
export default GameRequest;
