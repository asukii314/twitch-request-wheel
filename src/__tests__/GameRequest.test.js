import {ActivityStatus} from '../ChatActivity';
import {createRenderer} from 'react-test-renderer/shallow';
import GameRequest from '../components/GameRequest';
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';

// forces react-tooltip to reuse the same uuids
// to avoid breaking snapshots each time :P
jest.mock('crypto', () => ({
    randomBytes: (num: number) => new Array(num).fill(0),
}));

describe('GameRequest', () => {
    const propsQuip = {
        gameName: 'Quiplash (Standalone)',
        metadata: {
            name: 'Quiplash',
            chosen: true,
            locked: true,
            time: Date.now()-300000,
            username: 'Sir Goosewell'
        }
    };
    const propsSTI = {
    	gameName: 'Survive The Internet (Party Pack 4)',
    	metadata: {
    		name: 'Survive The Internet',
    		longName: 'Survive The Internet (Party Pack 4)',
    		partyPack: 'Party Pack 4',
    		'Min players': 3,
    		'Max players': 8,
    		Variants: ['survive', 'survive the internet', 'sti'],
    		username: 'dcpesses',
    		time: Date.now()-120000,
    		locked: false,
    		chosen: false
    	}
    };
    describe('deleteRequest', () => {
        test('should call props.onDelete', () => {
            const component = new GameRequest({
                onDelete: jest.fn().mockImplementation(()=>{})
            });
            jest.spyOn(component.props, 'onDelete').mockImplementation(()=>{});
            component.deleteRequest();
            expect(component.props.onDelete).toHaveBeenCalledTimes(1);
        });
    });
    describe('getFormattedTimeDiff', () => {
        test('should return the expected response', async () => {
            let component = new GameRequest({
                metadata: {
                    chosen: true,
                    locked: true,
                    time: Date.now(),
                    username: 'Sir Goosewell'
                }
            });
            expect(component.getFormattedTimeDiff(Date.now())).toBe('just now');
            expect(component.getFormattedTimeDiff(Date.now()-(10 * 60000))).toBe('10 minutes ago');
            expect(component.getFormattedTimeDiff(Date.now()-(60 * 60000))).toBe('1 hour ago');
            expect(component.getFormattedTimeDiff(Date.now()-(24 * 60 * 60000))).toBe('1 day ago');
        });
    });

    describe('getTooltipContent', () => {
        test('should call getTooltipActivityStatusClassName', () => {
            const component = new GameRequest({
                metadata: {
                    chosen: true,
                    locked: true,
                    time: Date.now(),
                    username: 'Sir Goosewell'
                }
            });
            jest.spyOn(component, 'getTooltipActivityStatusClassName').mockReturnValue('active');
            let container = component.getTooltipContent();
            expect(component.getTooltipActivityStatusClassName).toHaveBeenCalledTimes(1);
            expect(container).toMatchSnapshot();
        });
    });
    describe('getTooltipActivityStatusClassName', () => {
        test('should return the correct class names', () => {
            const component = new GameRequest();
            expect(component.getTooltipActivityStatusClassName()).toBe('');
            component.state.activityStatus = ActivityStatus.ACTIVE;
            expect(component.getTooltipActivityStatusClassName()).toBe('active');
            component.state.activityStatus = ActivityStatus.DISCONNECTED;
            expect(component.getTooltipActivityStatusClassName()).toBe('disconnected');
            component.state.activityStatus = ActivityStatus.IDLE;
            expect(component.getTooltipActivityStatusClassName()).toBe('idle');
        });
    });
    describe('toggleLock', () => {
        test('should call props.toggleLock', () => {
            const component = new GameRequest({
                toggleLock: jest.fn().mockImplementation(()=>{}),
                gameName: 'Job Job'
            });
            jest.spyOn(component.props, 'toggleLock').mockImplementation(()=>{});
            component.toggleLock();
            expect(component.props.toggleLock).toHaveBeenCalledTimes(1);
        });
    });
    describe('updateStatus', () => {
        test('should throw an error if caught', async () => { // part of a new "catch and release" program
            let component = new GameRequest({
                getActivity: jest.fn().mockRejectedValue(new Error('error stub')),
                metadata: {
                    chosen: true,
                    locked: true,
                    time: Date.now(),
                    username: 'Sir Goosewell'
                }
            });
            jest.spyOn(component, 'setState').mockImplementation(() => {});
            try {
                await component.updateStatus();
            } catch (e) {
                expect(e.message).toBe('error stub'); //eslint-disable-line jest/no-conditional-expect
            }

            expect(component.setState).not.toHaveBeenCalled();
        });
    });

    describe('render', () => {
        test('should render component using quiplash data', async () => {
            let {container} = render(<GameRequest {...propsQuip} />);
            try {
                await waitFor(() => screen.getByText(/Quiplash/i));
                const element = screen.getByText(/Quiplash/i);
                expect(element).toBeInTheDocument();
            } catch (e) {
                throw e;
            }
            expect(container).toMatchSnapshot();
        });
        test('should render component using survive the internet data', () => {
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<GameRequest {...propsSTI} />);
            let instance = shallowRenderer.getMountedInstance();
            instance.setState({
                activityStatus: ActivityStatus.ACTIVE,
                timeDiff: '2 minutes ago'
            });
            expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
            instance.setState({
                activityStatus: ActivityStatus.IDLE,
                timeDiff: '3 minutes ago'
            });
            expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
            instance.setState({
                activityStatus: ActivityStatus.DISCONNECTED,
                timeDiff: '4 minutes ago'
            });
            expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
            shallowRenderer.unmount();
        });
        test('should render component using survive the internet data and updated status info', async () => {
            let props = Object.assign({}, propsSTI, {
                getActivity: jest.fn()
                    .mockResolvedValueOnce(3)
                    .mockResolvedValueOnce(2)
                    .mockResolvedValueOnce(1),
                onDelete: jest.fn(),
                toggleLock: jest.fn()
            })
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<GameRequest {...props} />);
            let instance = shallowRenderer.getMountedInstance();
            instance.setState({
                activityStatus: ActivityStatus.IDLE,
                timeDiff: '2 minutes ago'
            });
            let component = shallowRenderer.getRenderOutput();
            await component.props.onEnter();
            expect(props.getActivity).toHaveBeenCalledTimes(1);

            /*
            // DISABLED FOR BOOTSTRAP TRANSITION
            // Get the children of the element div.options
            let optionsChildren = component.props.children.props.children.props.children.props.children;
            optionsChildren[0].props.onClick();
            optionsChildren[0].props.onClick();
            optionsChildren[1].props.onClick();
            expect(props.toggleLock).toHaveBeenCalledTimes(2);
            expect(props.onDelete).toHaveBeenCalledTimes(1);
            */
            shallowRenderer.unmount();
        });
    });
});
