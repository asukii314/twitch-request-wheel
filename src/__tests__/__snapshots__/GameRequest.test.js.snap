// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameRequest getTooltipContent should call getTooltipActivityStatusClassName 1`] = `
<Tooltip
  id="game-request-tooltip"
  placement="right"
>
  <span
    className="tooltipText"
  >
    Requested 
    0
     by 
    Sir Goosewell
    <span
      className="status active"
    />
  </span>
</Tooltip>
`;

exports[`GameRequest render should render component using quiplash data 1`] = `
<div>
  <div
    class="game-request-wrapper fade-in"
  >
    <div
      class="game-request chosen"
      id="baseDiv"
    >
      <div
        class="game-request-body"
      >
        <div>
          <b>
            Quiplash
          </b>
           
          <small />
        </div>
        <div
          class="options"
        >
          <img
            alt="lock"
            class="lock locked"
            src="file-stub"
          />
          <button
            class="deleteButton"
            type="button"
          >
            X
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GameRequest render should render component using survive the internet data 1`] = `
<OverlayTrigger
  defaultShow={false}
  onEnter={[Function]}
  overlay={[Function]}
  placement="left"
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <div
    className="game-request-wrapper fade-in"
  >
    <div
      className="game-request pending"
      id="baseDiv"
    >
      <div
        className="game-request-body"
      >
        <div>
          <b>
            Survive The Internet
          </b>
           
          <small>
            (Pack 4)
          </small>
        </div>
        <div
          className="options"
        >
          <img
            alt="lock"
            className="lock unlocked"
            onClick={[Function]}
            src="file-stub"
          />
          <button
            className="deleteButton"
            onClick={[Function]}
            type="button"
          >
            X
          </button>
        </div>
      </div>
    </div>
  </div>
</OverlayTrigger>
`;

exports[`GameRequest render should render component using survive the internet data 2`] = `
<OverlayTrigger
  defaultShow={false}
  onEnter={[Function]}
  overlay={[Function]}
  placement="left"
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <div
    className="game-request-wrapper fade-in"
  >
    <div
      className="game-request pending"
      id="baseDiv"
    >
      <div
        className="game-request-body"
      >
        <div>
          <b>
            Survive The Internet
          </b>
           
          <small>
            (Pack 4)
          </small>
        </div>
        <div
          className="options"
        >
          <img
            alt="lock"
            className="lock unlocked"
            onClick={[Function]}
            src="file-stub"
          />
          <button
            className="deleteButton"
            onClick={[Function]}
            type="button"
          >
            X
          </button>
        </div>
      </div>
    </div>
  </div>
</OverlayTrigger>
`;

exports[`GameRequest render should render component using survive the internet data 3`] = `
<OverlayTrigger
  defaultShow={false}
  onEnter={[Function]}
  overlay={[Function]}
  placement="left"
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <div
    className="game-request-wrapper fade-in"
  >
    <div
      className="game-request pending"
      id="baseDiv"
    >
      <div
        className="game-request-body"
      >
        <div>
          <b>
            Survive The Internet
          </b>
           
          <small>
            (Pack 4)
          </small>
        </div>
        <div
          className="options"
        >
          <img
            alt="lock"
            className="lock unlocked"
            onClick={[Function]}
            src="file-stub"
          />
          <button
            className="deleteButton"
            onClick={[Function]}
            type="button"
          >
            X
          </button>
        </div>
      </div>
    </div>
  </div>
</OverlayTrigger>
`;
